File: ios/ZweiCheck/APIClient.swift
import Foundation

final class APIClient {
    private let baseURL = URL(string: "https://zweicheck.kamilunavo.com")!
    private let cookieName = "zc_session"
    private let cookieStorage = HTTPCookieStorage()
    private let keychain = KeychainSessionStore()
    private var sessionToken: String?

    private lazy var session: URLSession = {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.httpCookieStorage = cookieStorage
        configuration.httpShouldSetCookies = true
        configuration.httpCookieAcceptPolicy = .always
        configuration.timeoutIntervalForRequest = 30
        configuration.timeoutIntervalForResource = 60
        return URLSession(configuration: configuration)
    }()

    init() {
        sessionToken = keychain.load()
        restoreSessionCookie()
    }

    func me() async throws -> APIUser { try await request("/api/auth/me", as: AuthEnvelope.self).user }

    func login(email: String, password: String) async throws -> APIUser {
        let envelope: AuthEnvelope = try await request("/api/auth/login", method: "POST", json: ["email": email, "password": password])
        persistSessionCookie()
        return envelope.user
    }

    func register(name: String, email: String, password: String) async throws -> APIUser {
        let envelope: AuthEnvelope = try await request("/api/auth/register", method: "POST", json: ["name": name, "email": email, "password": password])
        persistSessionCookie()
        return envelope.user
    }

    func logout() async {
        _ = try? await requestData("/api/auth/logout", method: "POST")
        clearSessionCredentials()
    }

    func resendVerification() async throws { let _: VerificationResult = try await request("/api/auth/resend-verification", method: "POST") }

    func checks() async throws -> [CheckItem] { try await request("/api/checks", as: ChecksEnvelope.self).checks }
    func trustRouting() async throws -> TrustRoutingEnvelope { try await request("/api/trust-routing", as: TrustRoutingEnvelope.self) }
    func activities() async throws -> ActivitiesEnvelope { try await request("/api/activities?limit=50", as: ActivitiesEnvelope.self) }

    func markActivityRead(id: String) async throws -> ActivityItem {
        try await request("/api/activities/\(id)/read", method: "PATCH", as: ActivityEnvelope.self).activity
    }

    func markAllActivitiesRead() async throws {
        _ = try await requestData("/api/activities/read-all", method: "POST")
    }

    func registerNativePush(token: String, environment: String) async throws {
        _ = try await requestData(
            "/api/push/native/tokens",
            method: "POST",
            json: ["token": token, "environment": environment]
        )
    }

    func unregisterNativePush(token: String, environment: String) async throws {
        _ = try await requestData(
            "/api/push/native/tokens",
            method: "DELETE",
            json: ["token": token, "environment": environment]
        )
    }

    func updatePresence(status: String, durationMinutes: Int?) async throws {
        var body: [String: Any] = ["status": status]
        if let durationMinutes { body["durationMinutes"] = durationMinutes }
        _ = try await requestData("/api/trust-routing/presence", method: "PUT", json: body)
    }

    func invite(email: String) async throws -> InvitationResult {
        try await request("/api/invitations", method: "POST", json: ["email": email], as: InvitationResult.self)
    }

    func acceptInvitation(code: String) async throws {
        let _: AcceptInvitationResult = try await request("/api/invitations/accept", method: "POST", json: ["code": code])
    }

    func createCheck(
        reviewerID: String,
        category: CheckCategory,
        description: String,
        amount: String?,
        images: [UploadImage]
    ) async throws -> CheckItem {
        var fields = [
            "reviewerId": reviewerID,
            "category": category.rawValue,
            "description": description,
            "urgency": "none"
        ]
        if let amount, !amount.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { fields["amount"] = amount }

        let boundary = "ZweiCheck-\(UUID().uuidString)"
        var body = Data()
        for (name, value) in fields {
            appendField(name: name, value: value, boundary: boundary, to: &body)
        }
        for image in images.prefix(3) {
            appendFile(fieldName: "images", image: image, boundary: boundary, to: &body)
        }
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)

        var request = URLRequest(url: baseURL.appending(path: "/api/checks"))
        request.httpMethod = "POST"
        request.httpBody = body
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        let data = try await perform(request)
        return try decode(CheckEnvelope.self, from: data).check
    }

    func respond(checkID: String, recommendation: Recommendation, note: String) async throws -> CheckItem {
        try await request("/api/checks/\(checkID)/respond", method: "POST", json: ["recommendation": recommendation.rawValue, "note": note], as: CheckEnvelope.self).check
    }

    func close(checkID: String) async throws { _ = try await requestData("/api/checks/\(checkID)/close", method: "POST") }

    func deleteAccount(password: String) async throws {
        _ = try await requestData("/api/account", method: "DELETE", json: ["password": password])
        clearSessionCredentials()
    }

    private func request<T: Decodable>(_ path: String, method: String = "GET", json: [String: Any]? = nil, as type: T.Type = T.self) async throws -> T {
        let data = try await requestData(path, method: method, json: json)
        return try decode(T.self, from: data)
    }

    private func requestData(_ path: String, method: String = "GET", json: [String: Any]? = nil) async throws -> Data {
        var request = URLRequest(url: baseURL.appending(path: path))
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let json {
            request.httpBody = try JSONSerialization.data(withJSONObject: json)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        return try await perform(request)
    }

    private func perform(_ originalRequest: URLRequest) async throws -> Data {
        var request = originalRequest

        // URLSession's cookie handling with an ephemeral/custom cookie store isn't reliable
        // enough for our HttpOnly server session. Attach the persisted token explicitly.
        if request.value(forHTTPHeaderField: "Cookie") == nil,
           let token = sessionToken ?? keychain.load(),
           !token.isEmpty {
            sessionToken = token
            request.setValue("\(cookieName)=\(token)", forHTTPHeaderField: "Cookie")
        }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw APIClientError.message("Keine gültige Serverantwort.")
        }

        captureSessionCookie(from: http, requestURL: request.url)

        guard (200..<300).contains(http.statusCode) else {
            let serverMessage = (try? decode(APIErrorEnvelope.self, from: data))?.error
                ?? "ZweiCheck konnte die Anfrage nicht abschließen."
            if http.statusCode == 401 {
                clearSessionCredentials()
                throw APIClientError.unauthorized(serverMessage)
            }
            throw APIClientError.message(serverMessage)
        }
        return data
    }

    private func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
        do { return try JSONDecoder().decode(type, from: data) }
        catch { throw APIClientError.message("Die Serverantwort konnte nicht gelesen werden.") }
    }

    private func appendField(name: String, value: String, boundary: String, to body: inout Data) {
        body.append("--\(boundary)\r\nContent-Disposition: form-data; name=\"\(name)\"\r\n\r\n\(value)\r\n".data(using: .utf8)!)
    }

    private func appendFile(fieldName: String, image: UploadImage, boundary: String, to body: inout Data) {
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"\(fieldName)\"; filename=\"\(image.fileName)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: \(image.mimeType)\r\n\r\n".data(using: .utf8)!)
        body.append(image.data)
        body.append("\r\n".data(using: .utf8)!)
    }

    private func captureSessionCookie(from response: HTTPURLResponse, requestURL: URL?) {
        guard let url = requestURL else { return }
        var headerFields: [String: String] = [:]
        for (key, value) in response.allHeaderFields {
            headerFields[String(describing: key)] = String(describing: value)
        }

        let responseCookies = HTTPCookie.cookies(withResponseHeaderFields: headerFields, for: url)
        guard let cookie = responseCookies.first(where: { $0.name == cookieName }) else { return }
        cookieStorage.setCookie(cookie)
        sessionToken = cookie.value
        try? keychain.save(cookie.value)
    }

    private func persistSessionCookie() {
        if let token = sessionToken, !token.isEmpty {
            try? keychain.save(token)
            return
        }
        guard let cookie = cookieStorage.cookies(for: baseURL)?.first(where: { $0.name == cookieName }) else { return }
        sessionToken = cookie.value
        try? keychain.save(cookie.value)
    }

    private func restoreSessionCookie() {
        guard let value = sessionToken ?? keychain.load(), let host = baseURL.host else { return }
        sessionToken = value
        let properties: [HTTPCookiePropertyKey: Any] = [
            .name: cookieName,
            .value: value,
            .domain: host,
            .path: "/",
            .secure: "TRUE",
            .expires: Date().addingTimeInterval(30 * 24 * 60 * 60)
        ]
        if let cookie = HTTPCookie(properties: properties) { cookieStorage.setCookie(cookie) }
    }

    private func clearSessionCredentials() {
        sessionToken = nil
        cookieStorage.removeCookies(since: .distantPast)
        keychain.clear()
    }
}

enum APIClientError: LocalizedError {
    case unauthorized(String)
    case message(String)

    var isUnauthorized: Bool {
        if case .unauthorized = self { return true }
        return false
    }

    var errorDescription: String? {
        switch self {
        case .unauthorized(let text), .message(let text): text
        }
    }
}

File: android/app/src/main/java/de/kamilunavo/zweicheck/ApiClient.kt
package de.kamilunavo.zweicheck

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

class ApiClient(context: Context) {
    private val sessionStore = SessionStore(context.applicationContext)
    private val cookieName = "zc_session"

    suspend fun me(): User = parseUser(request("/api/auth/me").getJSONObject("user"))

    suspend fun login(email: String, password: String): User = parseUser(
        request(
            "/api/auth/login",
            method = "POST",
            json = JSONObject().put("email", email).put("password", password),
        ).getJSONObject("user")
    )

    suspend fun register(name: String, email: String, password: String): User = parseUser(
        request(
            "/api/auth/register",
            method = "POST",
            json = JSONObject().put("name", name).put("email", email).put("password", password),
        ).getJSONObject("user")
    )

    suspend fun logout() {
        runCatching { request("/api/auth/logout", method = "POST") }
        sessionStore.clear()
    }

    suspend fun requestPasswordReset(email: String) {
        request("/api/auth/request-password-reset", "POST", JSONObject().put("email", email))
    }

    suspend fun checks(): List<CheckItem> {
        val array = request("/api/checks").getJSONArray("checks")
        return List(array.length()) { parseCheck(array.getJSONObject(it)) }
    }

    suspend fun check(id: String): CheckItem = parseCheck(request("/api/checks/$id").getJSONObject("check"))

    suspend fun trustRouting(): TrustRouting {
        val root = request("/api/trust-routing")
        val self = parsePresence(root.getJSONObject("self"))
        val array = root.getJSONArray("connections")
        return TrustRouting(self, List(array.length()) { parseConnection(array.getJSONObject(it)) })
    }

    suspend fun updatePresence(status: String, durationMinutes: Int?) {
        val body = JSONObject().put("status", status)
        durationMinutes?.let { body.put("durationMinutes", it) }
        request("/api/trust-routing/presence", "PUT", body)
    }

    suspend fun invite(email: String?): String {
        val body = JSONObject()
        email?.trim()?.takeIf { it.isNotEmpty() }?.let { body.put("email", it) }
        return request("/api/invitations", "POST", body).getString("code")
    }

    suspend fun acceptInvitation(code: String) {
        request("/api/invitations/accept", "POST", JSONObject().put("code", code.trim()))
    }

    suspend fun removeConnection(id: String) {
        request("/api/connections/$id", "DELETE")
    }

    suspend fun activities(filter: String = "all"): Activities {
        val root = request("/api/activities?limit=50&filter=${filter.encodeQueryValue()}")
        val array = root.getJSONArray("activities")
        return Activities(
            activities = List(array.length()) { parseActivity(array.getJSONObject(it)) },
            unreadCount = root.optInt("unreadCount", 0),
            nextBefore = root.nullableString("nextBefore"),
        )
    }

    suspend fun markActivityRead(id: String) {
        request("/api/activities/$id/read", "PATCH")
    }

    suspend fun markAllActivitiesRead() {
        request("/api/activities/read-all", "POST")
    }

    suspend fun respond(checkId: String, recommendation: Recommendation, note: String): CheckItem {
        val root = request(
            "/api/checks/$checkId/respond",
            "POST",
            JSONObject().put("recommendation", recommendation.wireValue).put("note", note),
        )
        return parseCheck(root.getJSONObject("check"))
    }

    suspend fun close(checkId: String) {
        request("/api/checks/$checkId/close", "POST")
    }

    suspend fun createCheck(
        reviewerId: String,
        fallbackReviewerId: String?,
        category: String,
        description: String,
        amount: String?,
        urgency: String,
        reminderMinutes: Int?,
        autoReroute: Boolean,
        images: List<UploadImage>,
    ): CheckItem {
        val fields = linkedMapOf(
            "reviewerId" to reviewerId,
            "category" to category,
            "description" to description,
            "urgency" to urgency,
        )
        amount?.trim()?.takeIf { it.isNotEmpty() }?.let { fields["amount"] = it }
        fallbackReviewerId?.takeIf { it.isNotBlank() }?.let { fields["fallbackReviewerId"] = it }
        reminderMinutes?.takeIf { it > 0 }?.let {
            fields["escalationReminderMinutes"] = it.toString()
            fields["escalationAutoReroute"] = autoReroute.toString()
        }

        val boundary = "ZweiCheck-${UUID.randomUUID()}"
        val out = ByteArrayOutputStream()
        fun write(value: String) = out.write(value.toByteArray(Charsets.UTF_8))
        for ((name, value) in fields) {
            write("--$boundary\r\n")
            write("Content-Disposition: form-data; name=\"$name\"\r\n\r\n")
            write("$value\r\n")
        }
        images.take(3).forEach { image ->
            write("--$boundary\r\n")
            write("Content-Disposition: form-data; name=\"images\"; filename=\"${image.fileName.safeFileName()}\"\r\n")
            write("Content-Type: ${image.mimeType}\r\n\r\n")
            out.write(image.bytes)
            write("\r\n")
        }
        write("--$boundary--\r\n")

        val root = execute(
            path = "/api/checks",
            method = "POST",
            contentType = "multipart/form-data; boundary=$boundary",
            body = out.toByteArray(),
        )
        return parseCheck(root.getJSONObject("check"))
    }

    suspend fun deleteAccount(password: String) {
        request("/api/account", "DELETE", JSONObject().put("password", password))
        sessionStore.clear()
    }

    private suspend fun request(path: String, method: String = "GET", json: JSONObject? = null): JSONObject =
        execute(
            path = path,
            method = method,
            contentType = if (json != null) "application/json" else null,
            body = json?.toString()?.toByteArray(Charsets.UTF_8),
        )

    private suspend fun execute(
        path: String,
        method: String,
        contentType: String?,
        body: ByteArray?,
    ): JSONObject = withContext(Dispatchers.IO) {
        val connection = (URL("${BuildConfig.API_BASE_URL}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 30_000
            readTimeout = 60_000
            useCaches = false
            setRequestProperty("Accept", "application/json")
            sessionStore.load()?.takeIf { it.isNotEmpty() }?.let {
                setRequestProperty("Cookie", "$cookieName=$it")
            }
            if (contentType != null) setRequestProperty("Content-Type", contentType)
            if (body != null) {
                doOutput = true
                setFixedLengthStreamingMode(body.size)
            }
        }

        try {
            if (body != null) connection.outputStream.use { it.write(body) }
            val status = connection.responseCode
            captureSessionCookie(connection)
            val bytes = (if (status in 200..299) connection.inputStream else connection.errorStream)
                ?.use { it.readBytes() }
                ?: ByteArray(0)
            val text = bytes.toString(Charsets.UTF_8)
            if (status !in 200..299) {
                if (status == 401) sessionStore.clear()
                val message = runCatching { JSONObject(text).optString("error") }
                    .getOrNull()?.takeIf { it.isNotBlank() }
                    ?: "ZweiCheck konnte die Anfrage nicht abschließen."
                throw ApiException(message, status)
            }
            if (text.isBlank()) JSONObject() else JSONObject(text)
        } finally {
            connection.disconnect()
        }
    }

    private fun captureSessionCookie(connection: HttpURLConnection) {
        val cookieHeaders = connection.headerFields.entries
            .filter { it.key?.equals("Set-Cookie", ignoreCase = true) == true }
            .flatMap { it.value.orEmpty() }
        val regex = Regex("(?:^|[;,]\\s*)${Regex.escape(cookieName)}=([^;]+)")
        cookieHeaders.firstNotNullOfOrNull { regex.find(it)?.groupValues?.getOrNull(1) }
            ?.takeIf { it.isNotBlank() }
            ?.let(sessionStore::save)
    }

    private fun parseUser(json: JSONObject) = User(
        id = json.getString("id"),
        email = json.getString("email"),
        name = json.getString("name"),
        emailVerified = json.optBoolean("emailVerified", false),
        createdAt = json.optString("createdAt"),
    )

    private fun parsePresence(json: JSONObject) = Presence(
        status = json.optString("status"),
        expiresAt = json.nullableString("expiresAt"),
        updatedAt = json.nullableString("updatedAt"),
    )

    private fun parseConnection(json: JSONObject): TrustConnection {
        val person = json.getJSONObject("person")
        return TrustConnection(
            connectionId = json.getString("connectionId"),
            person = Person(person.getString("id"), person.getString("name"), person.getString("email")),
            presence = parsePresence(json.getJSONObject("presence")),
        )
    }

    private fun parseAttachment(json: JSONObject) = Attachment(
        id = json.getString("id"),
        originalName = json.optString("originalName"),
        mimeType = json.optString("mimeType"),
        sizeBytes = json.optInt("sizeBytes", 0),
        url = json.optString("url"),
    )

    private fun parseCheck(json: JSONObject): CheckItem {
        val attachments = json.optJSONArray("attachments") ?: JSONArray()
        return CheckItem(
            id = json.getString("id"),
            requesterId = json.optString("requesterId"),
            requesterName = json.optString("requesterName"),
            reviewerId = json.optString("reviewerId"),
            reviewerName = json.optString("reviewerName"),
            category = json.optString("category"),
            description = json.optString("description"),
            amountCents = if (json.isNull("amountCents") || !json.has("amountCents")) null else json.optInt("amountCents"),
            urgency = json.optString("urgency"),
            status = json.optString("status"),
            recommendation = json.nullableString("recommendation"),
            responseNote = json.nullableString("responseNote"),
            respondedAt = json.nullableString("respondedAt"),
            closedAt = json.nullableString("closedAt"),
            createdAt = json.optString("createdAt"),
            updatedAt = json.optString("updatedAt"),
            attachments = List(attachments.length()) { parseAttachment(attachments.getJSONObject(it)) },
        )
    }

    private fun parseActivity(json: JSONObject) = ActivityItem(
        id = json.getString("id"),
        eventType = json.optString("eventType"),
        icon = json.optString("icon"),
        title = json.optString("title"),
        body = json.optString("body"),
        actorName = json.nullableString("actorName"),
        checkId = json.nullableString("checkId"),
        invitationId = json.nullableString("invitationId"),
        connectionId = json.nullableString("connectionId"),
        readAt = json.nullableString("readAt"),
        createdAt = json.optString("createdAt"),
    )

    private fun JSONObject.nullableString(key: String): String? =
        if (!has(key) || isNull(key)) null else optString(key).takeIf { it.isNotBlank() }

    private fun String.encodeQueryValue(): String = java.net.URLEncoder.encode(this, Charsets.UTF_8.name())
    private fun String.safeFileName(): String = replace(Regex("[^A-Za-z0-9._-]"), "_").take(100)
}
